### DISCUSSION TITLE
OpenMage (19.x and 20.x) appears incompatible with Amasty extensions

### DISCUSSION THREAD

**Discussion by durzel** [OpenMage/magento-lts#3898] [Questions & Answers]
### Preconditions (*)

1. 19.4.5
2. 20.5.0

### Steps to reproduce (*)

1. Migrate Magento 1 CE 1.9.4.5 install with one or more Amasty extensions installed to OpenMage (19.x or 20.x)
2. Check admin (Amasty Extensions -> Extensions & Notifications)
3. If extension(s) are missing, either all or some of the functionality they provide is absent.

### Expected result (*)

1. Expected - Amasty extensions reported as installed in backend:

![Screenshot 2024-03-25 at 11 52 37](https://github.com/OpenMage/magento-lts/assets/7250950/8d74d626-e06b-480b-ac5c-f986ff69057c)

2. Functionality provided by aforementioned extensions to be present and functional.

### Actual result (*)

![Screenshot 2024-03-25 at 11 55 34](https://github.com/OpenMage/magento-lts/assets/7250950/77bb166a-0582-489b-b489-daf3be0c5033)

1. No Amasty extensions appear to be installed.  
2. Critical functionality like "Shopping Cart Price Rules" appears to stop working.  Known active discount codes report "X coupon code is not valid".  Functionality like "Auto Add Promo Items" fails to work.

- I have tested OpenMage 19.4.5 and 20.5.0 installs (both via composer).
- I have also tried downgrading PHP from 8.2 (which I initially used) to 8, and also to 7.4, with no change.
- On same machine, an original unmigrated Magento 1 CE 1.9.4.5 install (on PHP 7.4) works as expected.
- The "Troubleshooter" Amasty provides is ineffective as the output appears to be the same on both a vanilla CE and OpenMage install.

I appreciate this isn't an Amasty support group, but was hoping someone might have experienced issues with these extensions are compatibility with OpenMage.  I was at the very least hoping that 19.x would be compatible, but OOB it isn't.

**Comment by durzel** [comment]
@ADDISON74 Thanks for the conversion, wasn't sure if it would be considered a OM bug.

**Comment by addison74** [comment]
I hope you followed those steps in a test environment and backup everything.

How did you upgrade?
Did you clear the cache?
Have you enable and checked the OpenMage log files?
Did you use the Magento Developer Mode?

Most likely there are errors related to the PHP version, which are not displayed. OpenMage works with PHP 8.3 inclusive, but we do not guarantee extensions.

**Comment by durzel** [comment]
Hi.

- Upgrade was done via Composer, following the [Manual Install guide](https://github.com/OpenMage/magento-lts?tab=readme-ov-file#composer).
- All carried out in multiple test environments, using a cut of the live database & website.
- Caches, etc cleared after each environment change (all caches are disabled in the backend).
- Built in log files (system.log, exception.log, etc) are enabled, I can't see any additional configuration for "OpenMage log files" ?

Have not tried Magento Developer Mode as yet, but will try that next.

To be clear - the extensions work and output is as expected using the same PHP version (7.4 - minimum requirement for OpenMage) with the standard Magento CE 1.9.4.5 install, and one that has been upgraded to OpenMage (19).  Because of this, it doesn't seem to be a PHP version problem to me?

**Reply by addison74** [reply]
Is there any free Amasty extension that doesn't work for you for testing with OpenMage 20.5.0 and PHP 8.3?

**Reply by durzel** [reply]
Unfortunately they are all paid extensions.  I don't think they offer any free Magento 1 ones.

**Comment by fballiano** [comment]
mmm weird, I have several stores all on OpenMage v20.something and all of them use some amasty extension with no problem. I had to patch them because they were breaking on php8, but all the fixes weren't really that hard.

this is a screenshot from one of my customer' store
<img width="311" alt="Screenshot 2024-03-25 alle 12 36 46" src="https://github.com/OpenMage/magento-lts/assets/909743/4269e526-2dfb-4e93-8cef-94bcea8ce07e">

**Reply by durzel** [reply]
That output looks different to what I'm seeing even on my live CE 1.9.4.5 install, possibly an older Amasty "Base"?  Still, yours acknowledges the existence of the extensions at least.

Can you speak to what changes you had to make for PHP 8.x compatibility?  Were they documented anywhere?

**Reply by fballiano** [reply]
on another customer' store I see this page isn't working properly (nobody ever noticed, me neither)

<img width="578" alt="Screenshot 2024-03-25 alle 12 41 03" src="https://github.com/OpenMage/magento-lts/assets/909743/49c46fbb-a504-4a72-ac40-7a4d4229823c">

but the extensions still work perfectly fine:
<img width="243" alt="Screenshot 2024-03-25 alle 12 41 57" src="https://github.com/OpenMage/magento-lts/assets/909743/a1da6c5b-beff-40e0-b7d6-89c80dbbc8f0">

**Reply by fballiano** [reply]
no I haven't documented the changes to support php8 because they were basic ones, refresh, page breaks because of a php error, fix it, didn't take more than a day (they're big and old stores) to fix all of them while migrating from php7.4 to 8.2

**Reply by fballiano** [reply]
also, I did those migrations like 2 years ago so I hardly remember

**Reply by alexh-swdev** [reply]
Hm, in general, Amasty extensions are working here (PHP8.3 OM 20.5.0)
![grafik](https://github.com/OpenMage/magento-lts/assets/18015530/f5161d9e-edc6-4756-901b-127542aa34f5)

Don't you have any log outputs, neither OM system/excpeption, even with the lowest logleve, nor from the webserver itself?

/Edit: It required some minor PHP related fixes

**Comment by durzel** [comment]
Amasty's "Extended Order Grid" appears to still work, as in I see a customised Sales -> Orders page, so _some_ functionality appears to be working.

"Special Promotions Pro" and "Auto Add Promo Items", which both interact with Shopping Cart Price Rules don't work, or not fully.  I still see the list of possible promo combinations that these modules add, but they don't take effect on the OpenMage store, either in the frontend or backend.

![Screenshot 2024-03-25 at 12 44 54](https://github.com/OpenMage/magento-lts/assets/7250950/613fcaea-0a96-4120-a9b9-8b25f7417bf2)

Additionally, a basic built-in "Percent of product price discount" rule with a fixed coupon code doesn't work (reports it is not valid).  So it would appear that whatever bug this is breaks Shopping Cart Price Rule functionality significantly, if not totally.

**Reply by fballiano** [reply]
mmm to me it smells that there were heavy customizations somewhere, like, do you have a app/code/local/Mage folder with overridden files from the core?

**Reply by durzel** [reply]
I thought so too, but the only thing in app/core/local/Mage is something the "Special Promotions Pro" module added.  Everything else is in app/code/local/Amasty

**Reply by addison74** [reply]
Enable the OM developer mode to catch the errors in the browser window.
Disable all Amasty extensions in /app/etc/modules/ then activate and test them one by one.

As @fballiano mentioned, most of the extensions need small adjustment in the code for PHP 8.x. Once you find the error message just search for it on the Internet.

**Reply by durzel** [reply]
Thanks.

For the moment I'll stick with OM 19.4.8 and PHP 7.4, and try and get it working on there before I get into PHP 8.x incompatibilities (and anything OM 20.x might change).

**Reply by fballiano** [reply]
all changes in OM20 are documented in the readme, apart the latest one where OM19 was frozen so everything went into OM20

**Reply by durzel** [reply]
Digging this up from the grave since I've only recently had chance to jump back on this, but after disabling Amasty's "Special Promotions Pro" module files (which are several - `Amasty_Base.xml,` `Amasty_PromoBannerLite.xml`, `Amasty_Promocopy.xml`, `Amasty_RulesPro.xml`, `Amasty_Rules.xml` and `Amasty_RuleTimeConditions.xml`) and basic promo codes start working again.

It would seem therefore that this module is not compatible with PHP 8, though I haven't seen any errors or warnings in any log files.  

Either way, not strictly an OpenMage problem.

**Reply by F1Red5** [reply]
@durzel you can provide patches here https://github.com/sreichel/openmage-patches or send me the extension (NDA) and i'll update them.

**Reply by durzel** [reply]
> @durzel you can provide patches here https://github.com/sreichel/openmage-patches or send my the extension (NDA) and i'll update them.

Thanks.  I'm not sure I follow though.. do you mean if I send you the Amasty extension(s) that are causing problems that you'll update them to work with PHP 8.2/OpenMage 20?

**Reply by F1Red5** [reply]
@durzel yes.

I'd provide patches to make the extensions work with php8.2/OM 20.

I'll not update the extension itself. This requires that the extension are installed via (recommended) composer.

**Reply by durzel** [reply]
> @durzel yes.
> 
> I'd provide patches to make the extensions work with php8.2/OM 20.
> 
> I'll not update the extension itself. This requires that the extension are installed via (recommended) composer.

Appreciate the offer.  I'm not sure how this would work if you're not updating the actual extension(s).  I don't know how you'd be able to provide patches or even work out where the Amasty extension is going wrong on OM 20 without insight and access to our systems.

I think we're probably going to just downgrade the server to PHP 7.4/OM 19 - since this is maintaining Magento 1 compatibility.

The fact this Amasty extension is just failing without any errors or warnings reported is making me wonder what else might be broken that I won't know about until I've actually deployed this server.

**Reply by sreichel** [reply]
I'd not update the extension itself. I'd offer a patch-file that could be used via "vendor-patches".

Requirement ... the extension needs to be installable via composer. (I know Amasty M1 extension don't have hit. You have to add composer.json and modman file yourself.)
